### PR TITLE
Experimental Unread Feed Shuffling (completely unsolicited feature)

### PR DESCRIPTION
**Pull request by clebens**
This is kind of a little personal project that I built on top of feedbin. Let me know if this is a feature you'd like in the main branch of the project.

I've long been frustrated with RSS readers getting flooded by extremely prolific feeds (or, in some cases, extremely popular feeds). As a result, posts from smaller blogs tend to get completely lost in the shuffle amongst bigger blogs. 

My goal is to come up with a good sorting solution for RSS feeds that:

* Provides a roughly chronological view
* Logically shuffles to reduce the flooding of prolific feeds
* Is user-data independent (thus avoiding flooding from very popular feeds) 

This is a work-in-progress roughly based on A* pathfinding algorithms. It takes each set of retrieved unread entries (seemingly 100 entries) and re-sorts according to a heuristic that uses, along with adjustable constants: 

* the number of entries already included from a selected entry's feed; and
* the time since the selected entry was published

I'm still playing around with it. It may be needlessly complex... (but I think its adjustability compensates for its complexity). In practice, it's already seeming to have a pretty dramatic impact on creating a more varied unread feed.

Note - I'm not really too much a Ruby / Rails expert. Let me know if my code:

* is inserted at an appropriate point
* even remotely comports with accepted Ruby / Rails styles, conventions, and idioms

Thanks!

### CODE REVIEW

**Comment by benubois** [comment]
Hi @clebens,

Love this idea! I did a [quick rewrite](https://github.com/feedbin/feedbin/commit/e5a74b6e3e187f525524565d03f569c010e3bff2) for style/formatting. Mostly this is just how I would write this sort of thing. Any other ideas you had for simplifying and clarifying this code would be welcome too.

My personal preference is to use [`Enumerable#each_with_object`](http://ruby-doc.org/core-2.2.2/Enumerable.html#method-i-each_with_object) over [`Hash::[]`](http://ruby-doc.org/core-2.2.2/Hash.html#method-c-5B-5D). I find the former more readable.

Another personal (maybe Ruby?) thing is around naming variables what they are. For example `feed_id` is preferable to `k` for me.

I'm happy with the location. If this was going to be used in multiple places maybe moving it/abstracting it would make more sense.

I'd be interested in merging this sooner rather than later to see how it looks with production data as long as you'd be up for tweaking as you see fit going forward.

One last thing to think about is what to call it in settings. **Shuffled** is an option but I'd like to give more of an explanation about what it did, like **Smart Shuffle**, **Weighted**, **Sort by Magic** [:wink: as an homage to Google Reader](http://www.gtricks.com/google-reader-tricks/what-is-sort-by-magic-in-google-reader/). A short explanation could fit below too.

File: app/controllers/entries_controller.rb [hunk 1]
```diff
@@ -32,6 +32,70 @@ def unread
     unread_entries = @user.unread_entries.select(:entry_id).page(params[:page]).sort_preference(@user.entry_sort)
     @entries = Entry.entries_with_feed(unread_entries, @user.entry_sort).entries_list
 
+    # Proper location for adjustments to unread entries?
+    # 
+    # structured as case to support future additional alternative
+    # post-query shuffle / sort options
+
+    case @user.entry_sort
+     when 'SHUFF'
```
Comment (benubois) [inline app/controllers/entries_controller.rb]: I'd rather see `if @user.entry_sort == "SHUFF"`.

File: app/controllers/entries_controller.rb [hunk 2]
```diff
@@ -32,6 +32,70 @@ def unread
...
+      # => ~ O(n) if n >> # of feeds ; O(n^2) if n ~= # of feeds
+
+      split_entries = {}
+    
+      @entries.map do |e| 
+        split_entries[e.feed_id] = [] unless split_entries[e.feed_id]
+        split_entries[e.feed_id].to_a.append e
+      end
+
+      feed_queue_lengths = Hash[ split_entries.keys.map { |k| split_entries[k].length } ]
```
Comment (benubois) [inline app/controllers/entries_controller.rb]: It looks like this isn't used anywhere. Can it be removed?

Comment (clebens) [inline app/controllers/entries_controller.rb]: Yes - I thought I would need it, but it ended up that my implementation only needed @entries.length.

File: app/controllers/entries_controller.rb (hunk 1 above)
Comment (clebens) [inline app/controllers/entries_controller.rb]: I agree - I should have changed it.

**Comment by clebens** [comment]
Sort by Magic is the thing I miss most about Google Reader!  Although - it did tend to get flooded by very popular feeds (and, undoubtedly, relied loads of unavailable social media data).

I like the idea of "Smart Shuffle" since it basically sounds like it's a feature similar to Google's "Sort by Magic," but not trying to replicate it.

All your changes look logical - I'll tuck that away as formatting / style tips.

Planned improvements:

* Abstract into a method
* Find a way to iterate over each feed fewer times within the loop (could have meaningful performance impact if a ton of users with a ton of feeds each simultaneously make a request)
* Play around with the constants and see if there could be a couple "sweet spots"

**Comment by benubois** [comment]
Closing for now. Feel free to submit a new PR or re-open if you ever want to resume work on this. Thanks!

**Pull request closed without merging**